### PR TITLE
feat: add fixes

### PR DESCRIPTION
**Pull request by r-k-b**
# Problem

On a huge codebase, it's a lot of work to find and fix the unqualified usages that this rule guards against.

# Solution

Supply a Fix that will qualify the usages.

~~Note that **the fix may have false-positives,** and add incorrect qualifications to functions or values that coincidentally have the same name as an imported function. Hence the "dirty" in the title.~~ Should be fine now!

Even then, implementing this dirty Fix and cleaning up the resulting errors was **much** faster than any alternative I could think of, on our 160k+ LoC project.

### CODE REVIEW

File: src/NoUnqualifiedFunctions.elm [hunk 1]
```diff
@@ -37,60 +55,297 @@ type alias Context =
...
+
+                    else if ignoreBecauseDifferentModule then
+                        oldContext
+
+                    else
+                        {- is not qualified; we should fix it(?)
+
+                           This function or value _might_ be shadowing the one
+                           we want to change!
+                           How do we tell that we shouldn't change it?
```
Comment (r-k-b) [inline src/NoUnqualifiedFunctions.elm]: [`withModuleNameLookupTable`](https://package.elm-lang.org/packages/jfmengels/elm-review/latest/Review-Rule#withModuleNameLookupTable) may be the answer...

Comment (r-k-b) [reply src/NoUnqualifiedFunctions.elm]: It was! Credit to Jeroen Engels for helping out: https://elmlang.slack.com/archives/C010RT4D1PT/p1686904504914439?thread_ts=1686904175.397819&cid=C010RT4D1PT

**Comment by galdiuz** [comment]
Having an automated fix sounds like a sound idea, even if it's not necessarily perfect. I'll do a review and some testing, and then I'll merge and publish.

File: src/NoUnqualifiedFunctions.elm [hunk 2]
```diff
@@ -37,60 +58,295 @@ type alias Context =
...
-                                )
+                                { errorIngredients =
+                                    ( { message = "Importing functions is not allowed. Use qualified names instead."
+                                      , details = [ detailsMessage oldContext args ]
+                                      }
+                                    , Node.range node
+                                      {- Should we provide a fix here for the import expression?
+
+                                         Or... do we let the `NoUnused` rule handle it? 😈
+                                      -}
```
Comment (galdiuz) [inline src/NoUnqualifiedFunctions.elm]: While having a fix locally in the module would be ideal, as you say `NoUnused.Variables.rule` does already handle that and that's good enough for me. Add a note in the readme about that, then remove this comment.

File: src/NoUnqualifiedFunctions.elm [hunk 3]
```diff
@@ -37,60 +58,295 @@ type alias Context =
...
+                                |> Dict.insert new.prefixWith new.errorIngredients
+                        , prefixesToApplyByName =
+                            accContext.prefixesToApplyByName
+                                |> Dict.insert new.searchFor
+                                    { matchModule = new.matchModule
+                                    , prefix = new.prefixWith
+                                    }
+                    }
+            in
+            -- could probably do both folds in a single pass...
```
Comment (galdiuz) [inline src/NoUnqualifiedFunctions.elm]: It's fine this way, might be more readable even. Remove this comment.

File: src/NoUnqualifiedFunctions.elm [hunk 4]
```diff
@@ -37,60 +58,295 @@ type alias Context =
...
+                    context.fixesToApplyByPrefix |> Dict.get prefix
+
+                fixes =
+                    case usageFixes of
+                        Just usages ->
+                            -- importFix ::
+                            usages
+
+                        Nothing ->
+                            [{- return only the importFix here -}]
```
Comment (galdiuz) [inline src/NoUnqualifiedFunctions.elm]: Since we're not fixing imports, clean up these comments.

**Review by galdiuz — CHANGES_REQUESTED** [review]
Looks good to me, and the fix worked in when I tested it. Clean up the comments in the code and add a note in the readme about the module providing automatic fixes while mentioning `NoUnused.Variables`, then I'll merge and publish.

File: src/NoUnqualifiedFunctions.elm (hunk 2 above)
Comment (r-k-b) [reply src/NoUnqualifiedFunctions.elm]: Sure! d69a29b1c19069a90de3456bb6fc4a6d0c390913

File: src/NoUnqualifiedFunctions.elm (hunk 3 above)
Comment (r-k-b) [reply src/NoUnqualifiedFunctions.elm]: e5fb33710c47821c51398c3747f0ba7e0fb7e0ca

File: src/NoUnqualifiedFunctions.elm (hunk 4 above)
Comment (r-k-b) [reply src/NoUnqualifiedFunctions.elm]: e5fb33710c47821c51398c3747f0ba7e0fb7e0ca

**Review by galdiuz — APPROVED** [review]

**Pull request merged**